viz/LaserLine.cpp

```diff
@@ -3,40 +3,137 @@
 
 using namespace vizkit3d;
 
-struct LaserLine::Data {
-    // Copy of the value given to updateDataIntern.
-    // Making a copy is required because of how OSG works
-    base::samples::LaserScan data;
-};
-
-
-LaserLine::LaserLine()
-    : p(new Data)
+LaserLine::LaserLine() : showRange(false),
+                        color(0, 255, 0, 255)
 {
+    vertices = new osg::Vec3Array();
+    colors = new ::osg::Vec4Array();
+    geometry = new osg::Geometry();
+    primitiveSet = new osg::DrawArrays();
 }
 
 LaserLine::~LaserLine()
 {
-    delete p;
 }
 
 osg::ref_ptr<osg::Node> LaserLine::createMainNode()
 {
-    return new osg::Geode();
+    /**
+     * set up the line style
+     */
+    geometry->getOrCreateStateSet()->setAttribute(new osg::LineWidth(3.5f), osg::StateAttribute::ON);
+    geometry->getOrCreateStateSet()->setMode(GL_LINE_SMOOTH, osg::StateAttribute::ON);
+    geometry->getOrCreateStateSet()->setMode(GL_BLEND, osg::StateAttribute::ON);
+    geometry->getOrCreateStateSet()->setMode(GL_LINE_SMOOTH_HINT, GL_NICEST);
+
+    // Add a blend function.
+    osg::ref_ptr<osg::BlendFunc> blend ( new osg::BlendFunc ( GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA ) );
+    geometry->getOrCreateStateSet()->setAttributeAndModes ( blend.get(), osg::StateAttribute::ON );
+
+    // Set the hint.
+    osg::ref_ptr<osg::Hint> hint ( new osg::Hint ( GL_LINE_SMOOTH_HINT, GL_NICEST ) );
+    geometry->getOrCreateStateSet()->setAttributeAndModes ( hint.get(), osg::StateAttribute::ON );
+
+
+    /**
+     * add the primitive set to geometry
+     */
+    geometry->addPrimitiveSet(primitiveSet);
+
+    /**
+     * set line colors
+     */
+    colors->push_back(osg::Vec4(color.red()  / 255.0, color.green()  / 255.0, color.blue()  / 255.0, color.alpha()  / 255.0));
+
+    osg::ref_ptr<osg::Geode> geode(new osg::Geode());
+    geode->addDrawable(geometry.get());
+
+    return geode.get();
 }
 
 void LaserLine::updateMainNode ( osg::Node* node )
 {
     osg::Geode* geode = static_cast<osg::Geode*>(node);
-    // Update the main node using the data in p->data
+
+    //laser scan depth values
+    std::vector<uint32_t> ranges = data.ranges;
+
+    double angle = data.start_angle; //start angle
+    double angleStep = data.angular_resolution; //angle step
+    double maxRange = data.maxRange / 1000.0; //convert millimeters to meters
+
+    vertices->clear();
+
+    if (showRange) {
```

Comment:
Remove this, please. There is already a plugin to display the ranges for a laser scanner, really no need to duplicate the functionality (at the cost of making this plugin more complicated).